{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "947756ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "from Bio import Entrez, SeqIO, pairwise2\n",
    "from IPython.display import Markdown\n",
    "import requests\n",
    "import time\n",
    "import json\n",
    "import re\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d73a560",
   "metadata": {},
   "source": [
    "#### A workflow to find the protein ID 'WP_XXXXXXXXX' of a specific gene (e.g. 'lysA')\n",
    "\n",
    "**Input:** a list of **genes** to find their corresponding protein IDs\n",
    "\n",
    "**Output:** a json of the inputed genes and their protein IDs (if found, else None)\n",
    "\n",
    "\n",
    "### Steps of the algorithm:\n",
    "1. Building a local list **protein_id_to_gene_map** (.json file) containing all the published genes of **C. glutamicum ATCC 13032** in RefSeq database.\n",
    "2. Searching through protein_id_to_gene_map for matches based on **gene name**\n",
    "\n",
    "If not found\n",
    "\n",
    "3. Collecting all the gene names from homologous C. glutamicum (12 strains available in KEGG)\n",
    "4. Searching through them for the gene name, when found, quering its amino acid sequence\n",
    "5. Searching through protein_id_to_gene_map for matches based on **AA sequence** homology\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7b8f160",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab0577f2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5f10b9d0",
   "metadata": {},
   "source": [
    "### Getting the gene data from the RefSeq database based on its protein ID\n",
    "\n",
    "Fetching every protein Id in **gene_to_reactions** and returning the data for its gene: gene_name, product, organism, sequence.\n",
    "\n",
    "The data is stored locally in **protein_id_to_gene_map** for future use, avoiding searching every time.\n",
    "\n",
    "### Serching the protein ID of a known gene \n",
    "The locally stored **protein_id_to_gene_map** are used to search for the **gene name** or its **sequence**\n",
    "\n",
    "### 1. Direct search through the collected list of annoctated genes:\n",
    "### 2. Searching through AA sequence homology:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b08d410d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SBML package 'layout' not supported by cobrapy, information is not parsed\n",
      "https://juser.fz-juelich.de/record/188973 does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n"
     ]
    }
   ],
   "source": [
    "from cobra.io import read_sbml_model\n",
    "\n",
    "model = read_sbml_model('iCGB21FR.xml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c3c0b54",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51dd3b39",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "728e46bf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "0b33db8b",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# 1. Direct search through the collected list of annoctated genes:Â¶\n",
    "\n",
    "def printBold(text1, text2=\"\"):\n",
    "    if len(text2)>0:\n",
    "        display(Markdown(f\"{text1} **{text2}**\"))\n",
    "    else:\n",
    "        display(Markdown(f\"{text1}\"))\n",
    "\n",
    "\n",
    "def getFullGeneName(protein_id):\n",
    "    for gene in model.genes:\n",
    "        if protein_id in gene.id:\n",
    "            return gene\n",
    "        \n",
    "def getProteinID(full_gene):\n",
    "    return \"WP_\" + re.search(r\"WP_(\\d{9})\", full_gene).group(1) if re.search(r\"WP_(\\d{9})\", full_gene) else None\n",
    "\n",
    "def fetchProteinID(protein_id, max_retries=50, delay_between_retries=1):\n",
    "    for attempt in range(max_retries):\n",
    "        try:\n",
    "            Entrez.email = \"h.bensaadi@gmail.com\"\n",
    "\n",
    "            handle = Entrez.efetch(db=\"protein\", id=protein_id, rettype=\"gb\", retmode=\"text\")\n",
    "            record = SeqIO.read(handle, \"genbank\")\n",
    "            handle.close()\n",
    "\n",
    "            gene_name, product, organism, sequence = None, None, None, None\n",
    "\n",
    "            for feature in record.features:\n",
    "                if 'gene' in feature.qualifiers:\n",
    "                    gene_name = feature.qualifiers['gene'][0]\n",
    "                if 'product' in feature.qualifiers:\n",
    "                    product = feature.qualifiers['product'][0]\n",
    "                if 'organism' in feature.qualifiers:\n",
    "                    organism = feature.qualifiers['organism'][0]\n",
    "\n",
    "            sequence = str(record.seq)\n",
    "\n",
    "            return protein_id, gene_name, product, organism, sequence\n",
    "\n",
    "        except Entrez.HTTPError as e:\n",
    "            if attempt < max_retries - 1:\n",
    "                print(f\"{protein_id}: retrying in {delay_between_retries} seconds...\")\n",
    "                time.sleep(delay_between_retries)\n",
    "            else:\n",
    "                print(f\"Max retries reached. Unable to fetch data for {protein_id}\")\n",
    "                raise\n",
    "        except Exception as e:\n",
    "            print(f\"Error: {e}\")\n",
    "            raise\n",
    "\n",
    "def loadProteinIDToGeneMap():\n",
    "    try:\n",
    "        with open(\"protein_id_to_gene_map.json\", \"r\") as file:\n",
    "            protein_id_to_gene_map = json.load(file)\n",
    "    except FileNotFoundError:\n",
    "        protein_id_to_gene_map = {}\n",
    "    \n",
    "    return protein_id_to_gene_map\n",
    "    \n",
    "def updateProteinIDToGeneMap(protein_id_to_gene_map):\n",
    "    with open(\"protein_id_to_gene_map.json\", \"w\") as file:\n",
    "        json.dump(protein_id_to_gene_map, file)\n",
    "           \n",
    "            \n",
    "def getNumberOfUniqueGenes():\n",
    "    \n",
    "    unique_protein_ids = []\n",
    "    \n",
    "    for gene in model.genes:\n",
    "        if getProteinID(gene.id) not in unique_protein_ids:\n",
    "            unique_protein_ids.append(getProteinID(gene.id))\n",
    "    return len(unique_protein_ids)\n",
    "    \n",
    "def mapProteinIDtoGene():\n",
    "    global protein_id_to_gene_map\n",
    "    \n",
    "    printBold(f\"Total number of unique genes in the model is: \", f\"{getNumberOfUniqueGenes()} genes\")\n",
    "    \n",
    "    protein_id_to_gene_map = loadProteinIDToGeneMap()\n",
    "\n",
    "    if len(protein_id_to_gene_map) < 1:\n",
    "        printBold(f\"\\nFetching RefSeq data for C. glutamicum ATCC 13032: \", f\"\")        \n",
    "\n",
    "        count_found_genes = 0\n",
    "        for gene in model.genes:\n",
    "            if \"WP_\" in gene.id:\n",
    "                protein_id = getProteinID(gene.id)\n",
    "\n",
    "                if protein_id not in protein_id_to_gene_map:\n",
    "                    protein_id, gene_name, product, organism, sequence = fetchProteinID(protein_id)\n",
    "                    protein_id_to_gene_map[protein_id] = [gene_name, product, organism, sequence]\n",
    "\n",
    "                    updateProteinIDToGeneMap(protein_id_to_gene_map)\n",
    "\n",
    "                    count_found_genes +=1\n",
    "\n",
    "                    print(f\"\\n{protein_id}, {gene_name}, {product}, {organism}, {sequence}\")\n",
    "\n",
    "        printBold(f\"Total number of fetched genes online is: \", f\"{count_found_genes}\")\n",
    "    \n",
    "    else:\n",
    "        printBold(f\"\\nRefSeq data from  C. glutamicum ATCC 13032 is already stored locally: \", f\"{len(protein_id_to_gene_map)+1} genes\")        \n",
    "\n",
    "# 2. Searching through AA sequence homology:\n",
    "\n",
    "def calculate_similarity(seq1, seq2):\n",
    "    alignments = pairwise2.align.globalxx(seq1, seq2, one_alignment_only=True)\n",
    "    if alignments:\n",
    "        alignment = alignments[0]\n",
    "        alignment_length = max(len(alignment[0]), len(alignment[1]))\n",
    "        similarity = alignment[2] / alignment_length\n",
    "        return similarity\n",
    "    return 0\n",
    "\n",
    "def searchGene(gene=None, reaction=None, seq=None, similarity_threshold=0.80):\n",
    "    global protein_id_to_gene_map\n",
    "    \n",
    "    matches = []\n",
    "    \n",
    "    for protein_id, data in protein_id_to_gene_map.items():\n",
    "        gene_name, product, organism, sequence = data\n",
    "        \n",
    "        if gene and gene_name and gene.lower() in gene_name.lower():\n",
    "            matches.append((protein_id, data))\n",
    "        \n",
    "        if reaction and product and reaction.lower() in product.lower():\n",
    "            matches.append((protein_id, data))\n",
    "        \n",
    "        if seq and sequence:\n",
    "            similarity = calculate_similarity(seq.lower(), sequence.lower())            \n",
    "            if similarity >= similarity_threshold:\n",
    "                matches.append((protein_id, f\"Similarity: {round(similarity*100, 2)}%\", data))\n",
    "    \n",
    "    if gene and len(matches)>0:\n",
    "        text_1_plural = \"matches\" if len(matches) != 1 else \"match\"\n",
    "        text_2_plural = \"were\" if len(matches) != 1 else \"was\"\n",
    "        \n",
    "        printBold(f\"Searching by gene name: \", f\"{len(matches)} {text_1_plural} {text_2_plural} found for {gene}\")\n",
    "    \n",
    "        return matches\n",
    "\n",
    "def getDataFromProtreinID(searched_protein_id):\n",
    "    protein_id_to_gene_map = loadMapFile()\n",
    "    for protein_id, data in protein_id_to_gene_map.items():\n",
    "        if searched_protein_id == protein_id:\n",
    "            return data\n",
    "\n",
    "def getGeneNameFromProteinID(searched_protein_id):\n",
    "    for gene, reactions in list(gene_to_reactions.items()):\n",
    "        if searched_protein_id == getProteinID(gene):\n",
    "            return gene\n",
    "    \n",
    "def getReactionsFromProteinID(searched_protein_id):\n",
    "    for gene, reactions in list(gene_to_reactions.items()):\n",
    "        if searched_protein_id == getProteinID(gene):\n",
    "            return reactions\n",
    "\n",
    "        \n",
    "def loadDataFromTaxonomySearch():\n",
    "    try:\n",
    "        with open(\"data_from_taxonomy_search.json\", \"r\") as file:\n",
    "            data_from_taxonomy_search = json.load(file)\n",
    "    except FileNotFoundError:\n",
    "        data_from_taxonomy_search = []\n",
    "    return data_from_taxonomy_search\n",
    "    \n",
    "def saveDataFromTaxonomySearch(data_from_taxonomy_search):\n",
    "    with open(\"data_from_taxonomy_search.json\", \"w\") as file:\n",
    "        json.dump(data_from_taxonomy_search, file)\n",
    "        \n",
    "def getDataBasedOnTaxonomy():\n",
    "    global data_from_taxonomy_search\n",
    "    taxonomy_codes = {\n",
    "        \"cgl\": \"Corynebacterium glutamicum ATCC 13032 (Kyowa Hakko)\",\n",
    "        \"cgb\": \"Corynebacterium glutamicum ATCC 13032 (Bielefeld)\",\n",
    "        \"cgu\": \"Corynebacterium glutamicum K051\",\n",
    "        \"cgt\": \"Corynebacterium glutamicum R\",\n",
    "        \"cgs\": \"Corynebacterium glutamicum SCgG1\",\n",
    "        \"cgg\": \"Corynebacterium glutamicum SCgG2\",\n",
    "        \"cgm\": \"Corynebacterium glutamicum MB001\",\n",
    "        \"cgj\": \"Corynebacterium glutamicum ATCC 21831\",\n",
    "        \"cgq\": \"Corynebacterium glutamicum AR1\",\n",
    "        \"cgx\": \"Corynebacterium glutamicum B253\"}\n",
    "\n",
    "    data_from_taxonomy_search = loadDataFromTaxonomySearch()\n",
    "    \n",
    "    if len(data_from_taxonomy_search) < 1:\n",
    "        print(f\"Fetching KEGG data for {len(taxonomy_codes)} C. glutamicum strain: \\n\")\n",
    "        \n",
    "        for taxonomy_code in taxonomy_codes:\n",
    "            search_url = f\"http://rest.kegg.jp/find/genes/{taxonomy_code}\"\n",
    "\n",
    "            try:\n",
    "                print(f\"Getting data for {taxonomy_codes[taxonomy_code]}\")\n",
    "                response = requests.get(search_url)\n",
    "                if response.status_code == 200:\n",
    "                    data_from_taxonomy_search.append(response.text.split('\\n'))\n",
    "                else:\n",
    "                    print(f\"Error: {response.status_code}\")\n",
    "            except requests.exceptions.RequestException as e:\n",
    "                print(f\"Request Exception: {e}\")\n",
    "\n",
    "        saveDataFromTaxonomySearch(data_from_taxonomy_search)\n",
    "    else:\n",
    "        printBold(f\"\\nKEGG data from other C. glutamicum strains is already stored locally: \", f\"{len(data_from_taxonomy_search)} strains\")\n",
    "            \n",
    "def getAminoAcidsSequence(gene_id):\n",
    "    response = requests.get(f\"http://rest.kegg.jp/get/{gene_id}\")\n",
    "    \n",
    "    if response.status_code == 200:        \n",
    "        lines = response.text.split('\\n')\n",
    "        amino_acid_sequence = \"\"\n",
    "        is_aa_sequence_section = False\n",
    "\n",
    "        for line in lines:\n",
    "            if line.startswith(\"AASEQ\"):\n",
    "                is_aa_sequence_section = True\n",
    "            elif line.startswith(\"NTSEQ\"):\n",
    "                is_aa_sequence_section = False\n",
    "            elif is_aa_sequence_section and line.strip(): \n",
    "                amino_acid_sequence += line.strip()\n",
    "        return amino_acid_sequence\n",
    "\n",
    "    else:\n",
    "        return f\"Error: {response.status_code}\"\n",
    "    \n",
    "    \n",
    "def getMatchedGenesByKEGGIdentifier():\n",
    "    global data_from_taxonomy_search\n",
    "    global genes_not_found\n",
    "    \n",
    "    found_homologous_genes = {}\n",
    "    \n",
    "    for gene_name_to_search in genes_not_found:\n",
    "        for data in data_from_taxonomy_search:\n",
    "            for entry in data:\n",
    "                if gene_name_to_search in entry:\n",
    "                    if gene_name_to_search not in found_homologous_genes:\n",
    "                        found_homologous_genes[gene_name_to_search] = [entry]\n",
    "                    else:\n",
    "                        found_homologous_genes[gene_name_to_search].append(entry)\n",
    "\n",
    "        print(f\"Homologous genes for {gene_name_to_search}: {len(found_homologous_genes[gene_name_to_search])} results\")\n",
    "\n",
    "    return found_homologous_genes\n",
    "\n",
    "\n",
    "def fetchAAsequenceOfHomologousGenes(found_homologous_genes):\n",
    "    \n",
    "    aa_sequence_results = {}\n",
    "    \n",
    "    print(\"\\nFetching AA sequence for the homologous genes: \\n\")\n",
    "\n",
    "    for gene_name, entries in list(found_homologous_genes.items()):\n",
    "        for entry in entries:\n",
    "            gene_id_kegg = entry.split(\"; \")[0].split(\"\\t\")[0]\n",
    "            gene_name_kegg = entry.split(\"; \")[0].split(\"\\t\")[1]\n",
    "            try:\n",
    "                gene_description_kegg = entry.split(\"; \")[1]\n",
    "            except:\n",
    "                pass\n",
    "\n",
    "            aa_seq = getAminoAcidsSequence(gene_id_kegg)\n",
    "\n",
    "            print(f\"{gene_name_kegg} ({gene_id_kegg}): {aa_seq[:25]}...\")\n",
    "\n",
    "            if gene_name not in aa_sequence_results:\n",
    "                aa_sequence_results[gene_name] = [{\"kegg_id\": gene_id_kegg, \"aa_seq\": aa_seq}]\n",
    "            else:\n",
    "                aa_sequence_results[gene_name].append({\"kegg_id\": gene_id_kegg, \"aa_seq\": aa_seq})\n",
    "\n",
    "    return aa_sequence_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "054bc351",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "Total number of unique genes in the model is:  **786 genes**"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "\n",
       "RefSeq data from  C. glutamicum ATCC 13032 is already stored locally:  **786 genes**"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "Searching by gene name:  **1 match was found for tkt**"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "\n",
       "KEGG data from other C. glutamicum strains is already stored locally:  **10 strains**"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Homologous genes for lysC: 2 results\n",
      "Homologous genes for ddh: 2 results\n",
      "Homologous genes for fbp: 11 results\n",
      "\n",
      "Fetching AA sequence for the homologous genes: \n",
      "\n",
      "lysC (cgb:cg0306): MALVVQKYGGSSLESAERIRNVAER...\n",
      "lysC (cgf:CGUA_00995): MALIVQKYGGSSLESAERIRAVAER...\n",
      "ddh (cgf:CGUA_11365): MAEAKIRAAIVGYGNLGRSAEKLLA...\n",
      "Obg family GTPase CgtA (ddh:Desde_3708): MFYDQAKIYVKGGDGGAGAVAFRRE...\n",
      "fbpA1 (cgf:CGUA_01390): MKLLRKLAAPAAIAIALGAGMSGVA...\n",
      "fbpA2 (cgf:CGUA_05650): MNRLPTRAGAVLTAVTVASLGVSGA...\n",
      "fbpA3 (cgf:CGUA_05655): MNTRTTVRYAAALVVAAVAAAPFPA...\n",
      "fbpA4 (cgf:CGUA_05670): MSPTPRRLAATAAALAVLANAFPAA...\n",
      "fbpA5 (cgf:CGUA_09355): MSRFTHRSLPSVLALGASAFLLAAC...\n",
      "fbpA6 (cgf:CGUA_10100): MRRSLTTVIATAVLAAAVTPAAAAQ...\n",
      "fbpC (cgf:CGUA_11260): MSTTRVLAVTLAAVLLAWPSAGAQA...\n",
      "fbpA7 (cgf:CGUA_12390): MRDTASSSRSPKNVPNLRRRGAALI...\n",
      "fbpB (cgf:CGUA_12395): MMSATAGLRRIHTTILALIVAVATA...\n",
      "Ferric transport system permease protein fbpB (hiq:CGSHiGG_03230): MNAKKLSIMHSAYFWIILSLLAFAL...\n",
      "fbpC (hiq:CGSHiGG_03235): MKFNKISLSVSTALLAAGLAVSGSA...\n",
      "\n",
      "Searching for AA sequence homolgy against genes from C. glutamicum ATCC 13032:\n",
      "\n",
      "Searching lysC -> cgb:cg0306: \n",
      "\n",
      "Searching lysC -> cgf:CGUA_00995: \n",
      "\n",
      "Searching ddh -> cgf:CGUA_11365: \n",
      "\n",
      "Searching ddh -> ddh:Desde_3708: \n",
      "\n"
     ]
    }
   ],
   "source": [
    "genes_to_find = [\"lysC\", \"ddh\", \"fbp\", \"tkt\"]\n",
    "found_genes_by_gene_name = []\n",
    "found_genes_by_aa_sequence = {}\n",
    "matches = []\n",
    "\n",
    "mapProteinIDtoGene()\n",
    "\n",
    "# 1. searching by gene_name\n",
    "for gene_name in genes_to_find:\n",
    "    match_results = searchGene(gene=gene_name)    \n",
    "    if match_results:\n",
    "        found_genes_by_gene_name.append(gene_name)\n",
    "        matches.append(match_results)\n",
    "\n",
    "# 2. searching by AA sequence homolgy\n",
    "\n",
    "getDataBasedOnTaxonomy()\n",
    "\n",
    "genes_not_found = list(set(genes_to_find) - set(found_genes_by_gene_name))\n",
    "\n",
    "found_homologous_genes = getMatchedGenesByKEGGIdentifier()\n",
    "aa_sequence_results = fetchAAsequenceOfHomologousGenes(found_homologous_genes)\n",
    "\n",
    "print(\"\\nSearching for AA sequence homolgy against genes from C. glutamicum ATCC 13032:\\n\")\n",
    "\n",
    "for gene_id, data in list(aa_sequence_results.items()):    \n",
    "    for element in data:\n",
    "        print(f'Searching {gene_id} -> {element[\"kegg_id\"]}: \\n')\n",
    "        match_results = searchGene(seq= element[\"aa_seq\"], similarity_threshold=0.65)\n",
    "        if match_results:\n",
    "            for match in match_results:\n",
    "                print(match)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7eda5171",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cc322f6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc45b8e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a2d7155",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19685251",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f582f9d0",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2de08aa2",
   "metadata": {},
   "source": [
    "## 3. Manual search of \"fbp\": "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55ccfd0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for metabolite in model.metabolites.query(\"D-Fructose 1-phosphate\", \"name\"):\n",
    "    print(f\"Metabolite ID for D-Fructose 1-phosphate is: {metabolite}\")\n",
    "    print(\"The reactions involving D-Fructose 1-phosphate  are:\")\n",
    "    for reaction in model.metabolites.f1p_c.reactions:\n",
    "        print(f\"\\t- {reaction.id} -> {reaction.reaction}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53aec3f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.FRUK "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2daaed4",
   "metadata": {},
   "source": [
    "# Summary:\n",
    "\n",
    "**Found through direct search:**\n",
    "\n",
    "- dapB = 'WP_011014794'\n",
    "- lysA = 'WP_011014180'\n",
    "- tkt = 'WP_011014456'\n",
    "\n",
    "**Found through AA sequence homology search:**\n",
    "\n",
    "- lysC = 'WP_003855724', 'Similarity: 100.0%'\n",
    "- ddh  = 'WP_011015446', 'Similarity: 66.49%'\n",
    "- pck  = 'WP_011013816', 'Similarity: 100.0%'\n",
    "- pyc  = 'WP_011013816', 'Similarity: 100.0%'\n",
    "- icd  = 'WP_011013800', 'Similarity: 100.0%'\n",
    "- hom  = 'WP_003854900', 'Similarity: 100.0%'\n",
    "\n",
    "**Found through manual search:**\n",
    "\n",
    "- fbp  = 'WP_011014765'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c3bcba5",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = {\n",
    "    'dapB':'WP_011014794',\n",
    "    'lysA': 'WP_011014180',\n",
    "    'tkt': 'WP_011014456',\n",
    "    'lysC': 'WP_003855724',\n",
    "    'ddh' : 'WP_011015446',\n",
    "    'pck': 'WP_011013816',\n",
    "    'pyc' : 'WP_011013816', \n",
    "    'icd': 'WP_011013800',\n",
    "    'hom' : 'WP_003854900', \n",
    "    'fbp' : 'WP_011014765'\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
