{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a784604e",
   "metadata": {},
   "source": [
    "## Replicating the results of Becker et al. 2011"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b7958c2",
   "metadata": {},
   "source": [
    "The issue with the model \"iCGB21FR.xml\" is that its authors have named the genes in the following format **\"lcl_{locus}_prot_{protein_id}\"**. An exemple name of the genes from the model would look like this: lcl_NC_006958_1_prot_WP_003859586_1_2138\n",
    "\n",
    "In which **\"NC_006958\"** is the locus represent the complete sequence of Corynebacterium glutamicum ATCC 13032, and **WP_XXXXXXXXX** represent the protein_id of its genes.\n",
    "\n",
    "\n",
    "**This results in adding an extra step of finding the gene name for the concerned protein ids.**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8a8c4cf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "%run gene_annotation.ipynb\n",
    "\n",
    "from colorama import init, Fore, Style\n",
    "from cobra import Model, Reaction\n",
    "from cobra.io import read_sbml_model\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "model = read_sbml_model('Files/iCGB21FR.xml')\n",
    "\n",
    "lys_exp_rxn = Reaction(\"LYSt3r\")\n",
    "lys_exp_rxn.name = \"L lysine transport out via proton antiport reversible\"\n",
    "lys_exp_rxn.add_metabolites({model.metabolites.h_e: -1, model.metabolites.lys__L_c: -1, model.metabolites.h_c: 1, model.metabolites.lys__L_e: 1})\n",
    "lys_exp_rxn.gene_reaction_rule = \"( Mbar_A2583 or Mbar_A3244 )\"\n",
    "lys_exp_rxn.lower_bound, lys_exp_rxn.upper_bound = 0, 1000\n",
    "\n",
    "model.add_reactions([lys_exp_rxn])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad43c832",
   "metadata": {},
   "source": [
    "### Annotating the genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "43fe46a6",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of unique genes in the model is: 786 genes\n",
      "\n",
      "RefSeq data from  C. glutamicum ATCC 13032 is already stored locally: 786 genes\n",
      "\n",
      "MATCH: dapB ->\n",
      "\n",
      "    ● WP_011014794 [4-hydroxy-tetrahydrodipicolinate reductase]\n",
      "\n",
      "MATCH: lysA ->\n",
      "\n",
      "    ● WP_011014180 [diaminopimelate decarboxylase]\n",
      "\n",
      "MATCH: tkt ->\n",
      "\n",
      "    ● WP_011014456 [transketolase]\n",
      "\n",
      "KEGG data from other C. glutamicum strains is already stored locally: 10 strains\n",
      "\n",
      "    Homologous genes for fbp: 12 results\n",
      "    Homologous genes for icd: 3 results\n",
      "    Homologous genes for pyc: 2 results\n",
      "    Homologous genes for hom: 6 results\n",
      "    Homologous genes for lysC: 4 results\n",
      "    Homologous genes for pck: 3 results\n",
      "\n",
      "Fetching AA sequence for the homologous genes: \n",
      "\n",
      "    ycgB (pvd:CFBP1590_0653): MTAREQKRQPISTGSEWTFELIQTY...\n",
      "    Fbp (cgu:WA5_0976): MNLKNPETPDRNLAMELVRVTEAAA...\n",
      "    fbpA1 (cgf:CGUA_01390): MKLLRKLAAPAAIAIALGAGMSGVA...\n",
      "    fbpA2 (cgf:CGUA_05650): MNRLPTRAGAVLTAVTVASLGVSGA...\n",
      "    fbpA3 (cgf:CGUA_05655): MNTRTTVRYAAALVVAAVAAAPFPA...\n",
      "    fbpA4 (cgf:CGUA_05670): MSPTPRRLAATAAALAVLANAFPAA...\n",
      "    fbpA5 (cgf:CGUA_09355): MSRFTHRSLPSVLALGASAFLLAAC...\n",
      "    fbpA6 (cgf:CGUA_10100): MRRSLTTVIATAVLAAAVTPAAAAQ...\n",
      "    fbpC (cgf:CGUA_11260): MSTTRVLAVTLAAVLLAWPSAGAQA...\n",
      "    fbpA7 (cgf:CGUA_12390): MRDTASSSRSPKNVPNLRRRGAALI...\n",
      "    fbpB (cgf:CGUA_12395): MMSATAGLRRIHTTILALIVAVATA...\n",
      "    fbpC (hiq:CGSHiGG_03235): MKFNKISLSVSTALLAAGLAVSGSA...\n",
      "    icd (cgb:cg0766): MAKIIWTRTDEAPLLATYSLKPVVE...\n",
      "    Icd (cgu:WA5_0634): MAKIIWTRTDEAPLLATYSLKPVVE...\n",
      "    icd (cgf:CGUA_03005): MSTIIWTRTDEAPLLATYSFKPVVE...\n",
      "    pyc (cgb:cg0791): MSTHTSSTLPAFKKILVANRGEIAV...\n",
      "    Pyc (cgu:WA5_0659): MSTHTSSTLPAFKKILVANRGEIAV...\n",
      "    hom (cgb:cg1337): MTSASAPSFNPGKGPGSAVGIALLG...\n",
      "    Hom (cgu:WA5_1136): MTSASAPSFNPGKGPGSAVGIALLG...\n",
      "    hom (cgf:CGUA_05360): MTATAQTSFNPGKGVGQPVGVAVLG...\n",
      "    obgE (rho:RHOM_07550): MFADRATIIIKSGKGGDGHVSFRRE...\n",
      "    obg (mhom:MLBD4_00730): MKFIDETNVIVIAGKGGDGIISFRR...\n",
      "    disrupted gene (stg:MGAS15252_0182): ...\n",
      "    lysC (cgb:cg0306): MALVVQKYGGSSLESAERIRNVAER...\n",
      "    LysC (cgu:WA5_0247): MALVVQKYGGSSLESAERIRNVAER...\n",
      "    lysC (cgf:CGUA_00995): MALIVQKYGGSSLESAERIRAVAER...\n",
      "    cgt (lcas:LYSCAS_22930): MTSVLLAGCTAPAPPPAPKQEAKAL...\n",
      "    pckA (lat:CGUJ_03595): MEKFDLEGSSRVYRNLSTSRLYEES...\n",
      "    Pck (cgu:WA5_2765): MTTAAIRGLQGEAPTKNKELLNWIA...\n",
      "    pckG (cgf:CGUA_12305): MTSAANTVRGLVGDAPTSNERLIAW...\n",
      "\n",
      "Searching for AA sequence homology against genes from C. glutamicum ATCC 13032:\n",
      "\n",
      "MATCH: Fbp (cgu:WA5_0976) ->\n",
      "\n",
      "    ● WP_003856830 (similarity: 100.0%) [class II fructose-bisphosphatase]: MNLKNPETPDRNLAMELVRVTEAAALASGRWVGRG...\n",
      "\n",
      "MATCH: fbpA5 (cgf:CGUA_09355) ->\n",
      "\n",
      "    ● WP_011013644 (similarity: 60.91%) [ABC transporter substrate-binding protein]: MSSKHPLKRTAVTVFALGASAALLVACSEPSEDVS...\n",
      "\n",
      "MATCH: icd (cgb:cg0766) ->\n",
      "\n",
      "    ● WP_011013800 (similarity: 100.0%) [NADP-dependent isocitrate dehydrogenase]: MAKIIWTRTDEAPLLATYSLKPVVEAFAATAGIEV...\n",
      "\n",
      "MATCH: Icd (cgu:WA5_0634) ->\n",
      "\n",
      "    ● WP_011013800 (similarity: 100.0%) [NADP-dependent isocitrate dehydrogenase]: MAKIIWTRTDEAPLLATYSLKPVVEAFAATAGIEV...\n",
      "\n",
      "MATCH: icd (cgf:CGUA_03005) ->\n",
      "\n",
      "    ● WP_011013800 (similarity: 65.8%) [NADP-dependent isocitrate dehydrogenase]: MAKIIWTRTDEAPLLATYSLKPVVEAFAATAGIEV...\n",
      "\n",
      "MATCH: pyc (cgb:cg0791) ->\n",
      "\n",
      "    ● WP_011013816 (similarity: 100.0%) [pyruvate carboxylase]: MSTHTSSTLPAFKKILVANRGEIAVRAFRAALETG...\n",
      "\n",
      "MATCH: Pyc (cgu:WA5_0659) ->\n",
      "\n",
      "    ● WP_011013816 (similarity: 100.0%) [pyruvate carboxylase]: MSTHTSSTLPAFKKILVANRGEIAVRAFRAALETG...\n",
      "\n",
      "MATCH: hom (cgb:cg1337) ->\n",
      "\n",
      "    ● WP_003854900 (similarity: 100.0%) [homoserine dehydrogenase]: MTSASAPSFNPGKGPGSAVGIALLGFGTVGTEVMR...\n",
      "\n",
      "MATCH: Hom (cgu:WA5_1136) ->\n",
      "\n",
      "    ● WP_003854900 (similarity: 100.0%) [homoserine dehydrogenase]: MTSASAPSFNPGKGPGSAVGIALLGFGTVGTEVMR...\n",
      "\n",
      "MATCH: hom (cgf:CGUA_05360) ->\n",
      "\n",
      "    ● WP_003854900 (similarity: 62.71%) [homoserine dehydrogenase]: MTSASAPSFNPGKGPGSAVGIALLGFGTVGTEVMR...\n",
      "\n",
      "MATCH: lysC (cgb:cg0306) ->\n",
      "\n",
      "    ● WP_003855724 (similarity: 100.0%) [aspartate kinase]: MALVVQKYGGSSLESAERIRNVAERIVATKKAGND...\n",
      "\n",
      "MATCH: LysC (cgu:WA5_0247) ->\n",
      "\n",
      "    ● WP_003855724 (similarity: 100.0%) [aspartate kinase]: MALVVQKYGGSSLESAERIRNVAERIVATKKAGND...\n",
      "\n",
      "MATCH: lysC (cgf:CGUA_00995) ->\n",
      "\n",
      "    ● WP_003855724 (similarity: 76.52%) [aspartate kinase]: MALVVQKYGGSSLESAERIRNVAERIVATKKAGND...\n",
      "\n",
      "MATCH: Pck (cgu:WA5_2765) ->\n",
      "\n",
      "    ● WP_011015446 (similarity: 99.67%) [phosphoenolpyruvate carboxykinase (GTP)]: MTTAAIRGLQGEAPTKNKELLNWIADAVELFQPEA...\n",
      "\n",
      "MATCH: pckG (cgf:CGUA_12305) ->\n",
      "\n",
      "    ● WP_011015446 (similarity: 66.49%) [phosphoenolpyruvate carboxykinase (GTP)]: MTTAAIRGLQGEAPTKNKELLNWIADAVELFQPEA...\n",
      "\n",
      "\n",
      "\n",
      "Summary:\n",
      "\n",
      "    ● dapB: WP_011014794\n",
      "    ● lysA: WP_011014180\n",
      "    ● tkt: WP_011014456\n",
      "    ● lysC: WP_003855724 (similarity 100.0%)\n",
      "    ● pck: WP_011015446 (similarity 99.67%)\n",
      "    ● hom: WP_003854900 (similarity 100.0%)\n",
      "    ● pyc: WP_011013816 (similarity 100.0%)\n",
      "    ● icd: WP_011013800 (similarity 100.0%)\n",
      "    ● fbp: WP_003856830 (similarity 100.0%)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'dapB': 'lcl_NC_006958_1_prot_WP_011014794_1_1871',\n",
       " 'lysA': 'lcl_NC_006958_1_prot_WP_011014180_1_1145',\n",
       " 'tkt': 'lcl_NC_006958_1_prot_WP_011014456_1_1529',\n",
       " 'lysC': 'lcl_NC_006958_1_prot_WP_003855724_1_250',\n",
       " 'pck': 'lcl_NC_006958_1_prot_WP_011015446_1_2735',\n",
       " 'hom': 'lcl_NC_006958_1_prot_WP_003854900_1_1147',\n",
       " 'pyc': 'lcl_NC_006958_1_prot_WP_011013816_1_666',\n",
       " 'icd': 'lcl_NC_006958_1_prot_WP_011013800_1_641',\n",
       " 'fbp': 'lcl_NC_006958_1_prot_WP_003856830_1_987'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%run gene_annotation.ipynb\n",
    "genes_to_find = [\"lysC\", \"pck\", \"dapB\", \"lysA\", \"hom\", \"pyc\", \"icd\", \"fbp\", \"tkt\"]\n",
    "found_matches = findGenes(genes_to_find, model)\n",
    "\n",
    "genes_map = printReport(found_matches)\n",
    "genes_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "844688ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = {\n",
    "    'dapB': 'lcl_NC_006958_1_prot_WP_011014794_1_1871',\n",
    "    'lysA': 'lcl_NC_006958_1_prot_WP_011014180_1_1145',\n",
    "    'tkt': 'lcl_NC_006958_1_prot_WP_011014456_1_1529',\n",
    "    'lysC': 'lcl_NC_006958_1_prot_WP_003855724_1_250',\n",
    "    'pck': 'lcl_NC_006958_1_prot_WP_011015446_1_2735',\n",
    "    'hom': 'lcl_NC_006958_1_prot_WP_003854900_1_1147',\n",
    "    'pyc': 'lcl_NC_006958_1_prot_WP_011013816_1_666',\n",
    "    'icd': 'lcl_NC_006958_1_prot_WP_011013800_1_641',\n",
    "    'fbp': 'lcl_NC_006958_1_prot_WP_003856830_1_987'}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed2e7f35",
   "metadata": {},
   "source": [
    "### Implementation of the genetic modifications\n",
    "\n",
    "@TODO: make it as table \n",
    "\n",
    "**LYS1:** \n",
    "- the mutation realsed **lysC** from the feedback inhibition => **skipped => gene regulation**\n",
    "**LYS2:**\n",
    "- having a second copy of **ddh** => the reaction coded by \"ddh\" (meso-diaminopimelate D-dehydrogenase: EC 1.4.1.16) does not exist in the model\n",
    "\n",
    "**LYS3:**\n",
    "- knockout of **pck** => **implemented by knocking out pck**\n",
    "\n",
    "**LYS4:**\n",
    "- over expression of **dapB** => **implemented by ↑ flux**\n",
    "**LYS5:**\n",
    "- additional copy of **lysA** => **implemented by ↑ flux**\n",
    "\n",
    "**LYS6:**\n",
    "- over expression of **lysC** => **implemented by ↑ flux**\n",
    "\n",
    "**LYS7:**\n",
    "- leaky variant of **hom** gene => **implemented by ↑ flux**\n",
    "\n",
    "**LYS8-9**\n",
    "- imporoving  the activity of **pyc** => **implemented by ↑ flux**\n",
    "\n",
    "**LYS10:**\n",
    "- overexpression of FBPase **fbp**  =>  **implemented by ↑ flux**\n",
    "\n",
    "**LYS11:**\n",
    "- overexpression  **tkt** => **implemented by ↑ flux**\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bb46500b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def optimizeModel(model, full_gene_name, fold_change=None, knock_out_gene=False):\n",
    "    \n",
    "    new_model = model.copy()\n",
    "    if fold_change is not None:\n",
    "        \n",
    "        for rxn in model.genes.get_by_id(full_gene_name).reactions:\n",
    "            reaction = model.reactions.get_by_id(rxn.id)\n",
    "            if reaction.lower_bound == 0:\n",
    "                reaction.lower_bound = 0.001\n",
    "            bounds = (fold_change * reaction.lower_bound, fold_change * reaction.upper_bound)\n",
    "            new_model.reactions.get_by_id(rxn.id).bounds = bounds\n",
    "\n",
    "    if knock_out_gene:\n",
    "        new_model.genes.get_by_id(full_gene_name).knock_out()\n",
    "\n",
    "    new_model.reactions.Growth.bounds = (0.259193805 , 1000)\n",
    "    new_model.objective = new_model.reactions.DAPDC\n",
    "    solution = new_model.optimize()\n",
    "    \n",
    "    return new_model, solution\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8b9af67f",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_2, solution_2 = optimizeModel(model, genes[\"icd\"], knock_out_gene=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9496cb6f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b3d64de",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9c7db1e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
